modules/light-clients/08-wasm/keeper/snapshotter.go

```diff
@@ -0,0 +1,115 @@
+package keeper
+
+import (
+	"io"
+
+	errorsmod "cosmossdk.io/errors"
+
+	snapshot "cosmossdk.io/store/snapshots/types"
+	storetypes "cosmossdk.io/store/types"
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/ibc-go/modules/light-clients/08-wasm/types"
+)
+
+var _ snapshot.ExtensionSnapshotter = &WasmSnapshotter{}
+
+// SnapshotFormat format 1 is just gzipped wasm byte code for each item payload. No protobuf envelope, no metadata.
+const SnapshotFormat = 1
+
+type WasmSnapshotter struct {
+	wasm *Keeper
+	cms  storetypes.MultiStore
+}
+
+func NewWasmSnapshotter(cms storetypes.MultiStore, wasm *Keeper) *WasmSnapshotter {
+	return &WasmSnapshotter{
+		wasm: wasm,
+		cms:  cms,
+	}
+}
+
+func (*WasmSnapshotter) SnapshotName() string {
+	return types.ModuleName
+}
+
+func (*WasmSnapshotter) SnapshotFormat() uint32 {
+	return SnapshotFormat
+}
+
+func (*WasmSnapshotter) SupportedFormats() []uint32 {
+	// If we support older formats, add them here and handle them in Restore
+	return []uint32{SnapshotFormat}
+}
+
+func (ws *WasmSnapshotter) SnapshotExtension(height uint64, payloadWriter snapshot.ExtensionPayloadWriter) error {
+	cacheMS, err := ws.cms.CacheMultiStoreWithVersion(int64(height))
+	if err != nil {
+		return err
+	}
+
+	ctx := sdk.NewContext(cacheMS, tmproto.Header{}, false, nil)
+
+	ws.wasm.IterateCode(ctx, func(code []byte) bool {
+		compressedWasm, err := types.GzipIt(code)
+		if err != nil {
+			return true
+		}
+
+		err = payloadWriter(compressedWasm)
+		return err != nil
+	})
+
+	return nil
+}
+
+func (ws *WasmSnapshotter) RestoreExtension(height uint64, format uint32, payloadReader snapshot.ExtensionPayloadReader) error {
+	if format == SnapshotFormat {
+		return ws.processAllItems(height, payloadReader, restoreV1, finalizeV1)
+	}
+	return snapshot.ErrUnknownFormat
+}
+
+func restoreV1(ctx sdk.Context, k *Keeper, compressedCode []byte) error {
+	if !types.IsGzip(compressedCode) {
+		return types.ErrInvalid.Wrap("not a gzip")
+	}
+	wasmCode, err := types.Uncompress(compressedCode, types.MaxWasmByteSize())
+	if err != nil {
+		return errorsmod.Wrap(errorsmod.Wrap(err, "failed to store contract"), err.Error())
+	}
+
+	_, err = k.wasmVM.StoreCode(wasmCode)
+	if err != nil {
+		return errorsmod.Wrap(errorsmod.Wrap(err, "failed to store contract"), err.Error())
+	}
+	return nil
+}
+
+func finalizeV1(ctx sdk.Context, k *Keeper) error {
+	return nil
+}
```

Comment:
I think we can delete if unused.

Reply:
seems like wasmd does the pinning here: https://github.com/CosmWasm/wasmd/blob/c6919e140b0267d18238b366606db23794e00035/x/wasm/keeper/snapshotter.go#L121 calls it https://github.com/CosmWasm/wasmd/blob/c6919e140b0267d18238b366606db23794e00035/x/wasm/keeper/keeper.go#L1023

instead of in `restoreV1`.

Reply:
My vote is just remove the finalization and do it in restore

Reply:
Aye, agreed! I will push some commits later. Some errors need to be cleaned up as well 👍